### PR TITLE
[GDBJIT] Make gdbjit thread-safe & bug fix

### PR DESCRIPTION
**Pull request by kbaladurin**
NotifyGdb::MethodPrepared method can be called from multiple threads simultaneously in this case gdbjit will work incorrectly as it uses global variable without synchronization. This patch reduces number of global variables and adds synchronization for them. Also it fixes DW_AT_comp_dir setting.

### CODE REVIEW

**Comment by kbaladurin** [comment]
cc @Dmitri-Botcharnikov @ayuckhulk

**Comment by kbaladurin** [comment]
@janvorli PTAL

File: src/vm/gdbjit.cpp [hunk 1]
```diff
@@ -661,15 +661,57 @@ void __attribute__((noinline)) __jit_debug_register_code() { __asm__(""); };
...
 struct jit_descriptor __jit_debug_descriptor = { 1, 0, 0, 0 };
+static CrstStatic JitDescriptorCrst;
 
 // END of GDB JIT interface
 
-/* Static data for .debug_str section */
-const char* DebugStrings[] = {
-  "CoreCLR", "" /* module name */, "" /* module path */
-};
+class DebugStringsCU
```
Comment (janvorli) [inline src/vm/gdbjit.cpp]: What does the "CU" mean?

File: src/vm/gdbjit.cpp [hunk 2]
```diff
@@ -661,15 +661,57 @@ void __attribute__((noinline)) __jit_debug_register_code() { __asm__(""); };
 /* Make sure to specify the version statically, because the
    debugger may check the version before we can set it.  */
 struct jit_descriptor __jit_debug_descriptor = { 1, 0, 0, 0 };
+static CrstStatic JitDescriptorCrst;
```
Comment (janvorli) [inline src/vm/gdbjit.cpp]: Coding convention for static variables in coreclr is `g_` prefix. Only static constants don't usually have this prefix. Could you please rename this instance to g_jitDescriptorCrst?

File: src/vm/gdbjit.cpp [hunk 3]
```diff
@@ -1785,48 +1827,23 @@ static int getNextPrologueIndex(int from, T &arr, int n)
     return -1;
 }
 
+static NewArrayHolder<WCHAR> WszModuleNames;
```
Comment (janvorli) [inline src/vm/gdbjit.cpp]: Could you please change the naming convention here to use `g_` prefix too?

File: src/vm/gdbjit.cpp [hunk 4]
```diff
@@ -2877,6 +2928,8 @@ void NotifyGdb::MethodPitched(MethodDesc* methodDescPtr)
     if (pCode == NULL)
         return;
 
+    CrstHolder _crst(&JitDescriptorCrst);
```
Comment (janvorli) [inline src/vm/gdbjit.cpp]: Naming convention - could you please remove the leading underscore from the name?

File: src/vm/gdbjit.cpp [hunk 5]
```diff
@@ -2547,21 +2597,25 @@ void NotifyGdb::OnMethodPrepared(MethodDesc* methodDescPtr)
...
 
-    /* Link into list */
-    jit_code_entry *head = __jit_debug_descriptor.first_entry;
-    __jit_debug_descriptor.first_entry = jit_symbols;
-    if (head != 0)
     {
-        jit_symbols->next_entry = head;
-        head->prev_entry = jit_symbols;
-    }
+        CrstHolder _crst(&JitDescriptorCrst);
```
Comment (janvorli) [inline src/vm/gdbjit.cpp]: Naming convention - could you please remove the leading underscore from the name?

File: src/vm/gdbjit.cpp (hunk 1 above)
Comment (kbaladurin) [reply src/vm/gdbjit.cpp]: "CU" means "Compile Unit". This class corresponds `DebugInfoCU` structure (`gdbjit.cpp:811`).

File: src/vm/gdbjit.cpp (hunk 3 above)
Comment (kbaladurin) [reply src/vm/gdbjit.cpp]: done

File: src/vm/gdbjit.cpp (hunk 5 above)
Comment (kbaladurin) [reply src/vm/gdbjit.cpp]: done

File: src/vm/gdbjit.cpp (hunk 4 above)
Comment (kbaladurin) [reply src/vm/gdbjit.cpp]: done

**Comment by kbaladurin** [comment]
@janvorli thank you for review. I've updated PR.

File: src/vm/gdbjit.cpp (hunk 2 above)
Comment (kbaladurin) [reply src/vm/gdbjit.cpp]: done

**Comment by kbaladurin** [comment]
@janvorli thank you!

**Pull request merged**